src/main/java/com/epam/aidial/core/upstream/UpstreamState.java

```diff
@@ -0,0 +1,69 @@
+package com.epam.aidial.core.upstream;
+
+import com.epam.aidial.core.config.Upstream;
+import com.epam.aidial.core.util.HttpStatus;
+import lombok.Getter;
+
+public class UpstreamState implements Comparable<UpstreamState> {
+
+    @Getter
+    private final Upstream upstream;
+    private final int errorsThreshold;
+
+    private static final long INITIAL_BACKOFF_DELAY_MS = 1000;
+    private static final int BACKOFF_BASE = 2;
+
+    /**
+     * Amount of 5xx errors from upstream
+     */
+    private int errorCount;
+    /**
+     * Timestamp in millis when upstream may be available
+     */
+    private long retryAfter = -1;
+
+    public UpstreamState(Upstream upstream, int errorsThreshold) {
+        this.upstream = upstream;
+        this.errorsThreshold = errorsThreshold;
+    }
+
+    /**
+     * Register upstream failure. Supported error codes are 429 and 5xx.
+     *
+     * @param status response status code from upstream
+     * @param retryAfterSeconds time in seconds when upstream may become available; only take into account with 429 status code
+     */
+    public synchronized void failed(HttpStatus status, long retryAfterSeconds) {
+        if (status == HttpStatus.TOO_MANY_REQUESTS) {
+            retryAfter = System.currentTimeMillis() + retryAfterSeconds * 1000;
```

Comment:
Is it guaranteed to be always non-negative?

Reply:
yes, check Retry-After header specification

Reply:
Long.parseLong could return negatives